libsql-sqlite3/src/vector.c

```diff
@@ -0,0 +1,581 @@
+  sqlite3_context *context,
+  Vector *v
+){
+  switch (v->type) {
+    case VECTOR_TYPE_FLOAT32:
+      vectorF32Deserialize(context, v);
+      break;
+    case VECTOR_TYPE_FLOAT64:
+      vectorF64Deserialize(context, v);
+      break;
+    default:
+      assert(0);
+  }
+}
+
+static void vectorSerialize(
+  sqlite3_context *context,
+  Vector *v
+){
+  if( v->dims==0 ) {
+    sqlite3_result_null(context);
+    return;
+  }
+  switch (v->type) {
+    case VECTOR_TYPE_FLOAT32:
+      vectorF32Serialize(context, v);
+      break;
+    case VECTOR_TYPE_FLOAT64:
+      vectorF64Serialize(context, v);
+      break;
+    default:
+      assert(0);
+  }
+}
+
+size_t vectorSerializeToBlob(Vector *p, unsigned char *blob, size_t blobSize){
+  switch (p->type) {
+    case VECTOR_TYPE_FLOAT32:
+      return vectorF32SerializeToBlob(p, blob, blobSize);
+    case VECTOR_TYPE_FLOAT64:
+      return vectorF64SerializeToBlob(p, blob, blobSize);
+    default:
+      assert(0);
+  }
+  return 0;
+}
+
+size_t vectorDeserializeFromBlob(Vector *p, const unsigned char *blob, size_t blobSize){
+  switch (p->type) {
+    case VECTOR_TYPE_FLOAT32:
+      return vectorF32DeserializeFromBlob(p, blob, blobSize);
+    case VECTOR_TYPE_FLOAT64:
+      
+    default:
+      assert(0);
+  }
+  return 0;
+}
+
+/**************************************************************************
+** Vector index cursor implementations
+****************************************************************************/
+
+/*
+** A VectorIdxCursor is a special cursor to perform vector index lookups.
+ */
+struct VectorIdxCursor {
+  sqlite3 *db;          /* Database connection */
+  DiskAnnIndex *index;   /* DiskANN index on disk */
+};
+
+/**
+** Parses a type string such as `FLOAT32(3)` and returns the number of dimensions.
+**
+** Returns -1 if the type string is not a valid vector type.
+**/
+static int parseVectorDims(const char *zType){
+  int dims = 0;
+  if( sqlite3_strnicmp(zType, "FLOAT32(", 8)==0 || sqlite3_strnicmp(zType, "FLOAT64(", 8)==0){
+    const char *z = zType + 8;
+    while( *z && *z!=')' ){
+      if( !sqlite3Isdigit(*z) ){
+        return -1;
+      }
+      dims = dims*10 + (*z - '0');
+      z++;
+    }
+    if( *z==0 ){
+      return -1;
+    }
+  } else {
+    return -1;
+  }
+  return dims;
+}
+
+int vectorIndexCreate(Parse *pParse, Index *pIdx, IdList *pUsing){
+  sqlite3 *db = pParse->db;
+  int nDistanceOps = 0;
+  Column *pCol;
+  Table *pTab;
+  char *zSql;
+  int nDims;
+  int rc;
+  int i;
+
+  assert( pUsing!= 0);
+
+  for( i=0; i<pUsing->nId; i++ ){
+    if( sqlite3_stricmp(pUsing->a[i].zName, "diskann_cosine_ops")==0 ){
+      nDistanceOps = VECTOR_DISTANCE_COS;
+      break;
+    }
+    sqlite3ErrorMsg(pParse, "Unknown indexing method: %s", pUsing->a[i].zName);
+    return -1;
+  }
+  zSql = sqlite3MPrintf(db, "CREATE TABLE IF NOT EXISTS %s_shadow (index_key INT, data BLOB)", pIdx->zName);
+  rc = sqlite3_exec(db, zSql, 0, 0, 0);
```

Comment:
I'm probably still not understanding something but it seems that `return -1` in line 393 and 399 will leave shadow table present despite index creation failing. Is this a problem? Probably nothing bad happens despite some junk being kept around. Creation of the index with the same name will just work because of `IF NOT EXISTIS` part in the SQL.